### ISSUE TITLE
Support for deleting keys

### ISSUE THREAD

**Issue by judepereira** [CleverTap/stormdb#20]

**Comment by AkshayS96** [comment]
@judepereira 
We can implement deletion using 2 steps:
1. remove the key from indexMap
2. add that key to an arraylist or write it to a file (for fault tolerance)
3. skip this key while compaction will be performed.

**Comment by judepereira** [comment]
What about during iteration? Will we have to check the indexMap at all times? Iteration performance might suffer.

**Comment by AkshayS96** [comment]
I think we check whether a key is present in index map or not at the time of iteration itself  here https://github.com/CleverTap/stormdb/blob/53edb09dace64e88e0a705877705d31053c96a3d/src/main/java/com/clevertap/stormdb/StormDB.java#L614

**Comment by judepereira** [comment]
The index map isn't checked. Only the bitset is checked.

**Comment by judepereira** [comment]
One good way to avoid checking the index map would be to "fake" reading them. by setting the keys directly to read in the `keysRead` bitset:
https://github.com/CleverTap/stormdb/blob/53edb09dace64e88e0a705877705d31053c96a3d/src/main/java/com/clevertap/stormdb/StormDB.java#L612

**Comment by AkshayS96** [comment]
yeah, we can do this. first, iterate through the deleted keys array list and set them to this bitset.

**Comment by judepereira** [comment]
Sounds like a plan :)

**Comment by judepereira** [comment]
What do you think is the best method for persistence of the deleted keys?

**Comment by AkshayS96** [comment]
I was thinking of keeping both the in-memory data structure (Map or ArrayList) and a file(in case of sudden shutdown for persistence);

**Comment by judepereira** [comment]
Sounds good :)
